File: frameworks/node/express/UnitTests/query.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import request from "supertest";

import { app, expected, run } from "./app.ts";

const SEARCH = ["page", "size", "status", "category", "sort", "q", "minPrice", "maxPrice"] as const;

// rb:test query.one
test("query.one: the page is echoed as an integer", async () => {
  const response = await request(app).get("/query/one").query({ page: String(run.page) });

  assert.deepEqual(response.body, expected.withEcho("items.small.json", { page: run.page }));
});

// rb:test query.many
test("query.many: eight values are echoed, the numbers as integers", async () => {
  const query = Object.fromEntries(SEARCH.map((name) => [name, String(run[name])]));

  const response = await request(app).get("/query/many").query(query);

  assert.deepEqual(response.body, expected.withEcho("items.small.json", Object.fromEntries(SEARCH.map((name) => [name, run[name]]))));
});

test("nothing checks a query value, so a missing page is written back as null", async () => {
  const response = await request(app).get("/query/one");

  assert.equal(response.status, 200);
  assert.equal(response.body.echo.page, null);
});

File: frameworks/go/net-http/UnitTests/query_test.go
package unittests

import "testing"

// rb:test query.one,query.many
func TestQuery(t *testing.T) {
	t.Run("query.one", func(t *testing.T) {
		assertOK(t, withEcho(t, "items.small.json", map[string]any{"page": 417.0}), get(t, "/query/one?page=417"))
	})

	t.Run("query.many", func(t *testing.T) {
		want := withEcho(t, "items.small.json", search())
		path := "/query/many?page=417&size=38&status=paid&category=garden&sort=created&q=alpha%20bravo&minPrice=1200&maxPrice=34000"
		assertOK(t, want, get(t, path))
	})
}

// search is query.many's eight values as the handler echoes them, the numbers as numbers.
func search() map[string]any {
	return map[string]any{
		"page": 417.0, "size": 38.0, "status": "paid", "category": "garden",
		"sort": "created", "q": "alpha bravo", "minPrice": 1200.0, "maxPrice": 34000.0,
	}
}
